Generic HPC Install Script

TimothyWillard opened the pull request:
### Describe your changes.

This adds a generic `hpc_install.sh` script which can reproducibly setup and install `flepiMoP` on both rockfish and longleaf. The script vaguely:

1. Figures out HPC specific variables and modules, currently only rockfish and longleaf are supported.
2. Loads sensitive credentials.
3. Sets up a `FLEPI_PATH` environment variable.
4. Sets up/updates a conda environment.
5. Ensures that the R/python versions of arrow are compatible. These checks are loose and not definitive.
6. Install custom R packages.
7. Set up environment variables common for use with `flepiMoP`.

Going to add a separate PR for documentation since that needs to be merged into `gitbook-documentation`. But user usage would look something like on rockfish:
```bash
wget https://raw.githubusercontent.com/HopkinsIDD/flepiMoP/refs/heads/GH-191/longleaf-batch-submission/build/hpc_install.sh
vim /scratch4/struelo1/flepimop-code/ext-twillard/slack_credentials.sh
chmod 600 /scratch4/struelo1/flepimop-code/ext-twillard/slack_credentials.sh
source hpc_install.sh rockfish
```
and on longleaf:
```bash
wget https://raw.githubusercontent.com/HopkinsIDD/flepiMoP/refs/heads/GH-191/longleaf-batch-submission/build/hpc_install.sh
vim /users/t/w/twillard/slack_credentials.sh
chmod 600 /users/t/w/twillard/slack_credentials.sh
source hpc_install.sh longleaf
```
And replacing the url with the appropriate one. Then keeping your environment up to date is as easy as:
```bash
source /scratch4/struelo1/flepimop-code/ext-twillard/flepiMoP/build/hpc_install.sh rockfish
```
on rockfish or:
```bash
source /users/t/w/twillard/flepiMoP/build/hpc_install.sh longleaf
```
on longleaf.

A big open question is how best to install packages. Right now it is installing `gempyor`, `flepiconfig`, `flepicommon`, and `inference` from GitHub rather than locally. Whereas I think installing locally would be preferred dev reasons, at least in the meantime.

### What does your pull request address? Tag relevant issues.

One of many steps required for GH-191. Should resolve GH-308

### Tag relevant team members.

@pearsonca, @shauntruelove, @MacdonaldJoshuaCaleb  

Edit: Fix `wget` url to use the "raw" file instead of the pretty version.

pearsonca commented on build/hpc_install.sh:
```diff
@@ -0,0 +1,157 @@
+elif [[ $1 == "rockfish" ]]; then
+    # Setup general purspose user variables needed for RockFish
+    USERDIR="/scratch4/struelo1/flepimop-code/$USER/"
+
```
and if we do, several of the $USERDIRs below can/must be eliminated

pearsonca replied:
could add creating some hpc-wide environmental variables to the longleaf-setup repo. does that make sense to pair with this?

lastly ... bit weird that we're doing install here in scratch. why not in $HOME? i get doing projects on scratch.

pearsonca replied:
per in-person conversation:
 - need to check the preferred location for *libraries* on longleaf & rockfish
 - maybe refer to that as $LIBDIR (or ACCLIBDIR or some such)
 - might want to move that as a generic variable to be set on the HPC, and if so - move that to the longleaf-setup directions (which could itself stand to be scriptified) and make that setup a prerequisite to this? (one downside to that would be other people on other HPCs wanting to use / modify this script - future problem?)

TimothyWillard replied:
At least on the longleaf side it looks like `/users` is similar to `$HOME`, the documentation states "Think of it as a capacity expansion to your home directory." However, I think maybe the project directory should be moved to `/work` since that's high throughput and designed for active jobs. So my take is:

- `flepiMoP` and `flepimop-env` stay in `/users`, especially for the conda env since that directory can get large and `$HOME` has some low and strict storage caps.
- Move the project directory to `/work` since that'll actually need throughput for the job.

I still need to dig up the rockfish documentation. Longleaf docs: https://help.rc.unc.edu/getting-started-on-longleaf/#main-directory-spaces.

pearsonca replied:
seems reasonable. just to double check:
 - library elements (all the conda environment setting, which includes flepimop-specific libraries) lives on /users/[something] - for a job, the idea (to be confirmed) is that stuff gets loaded from here approximately once. idea being: with the exception of very short running jobs, acceptably low overhead tax to have here instead of on work
 - project elements (all the run-specific inputs + outputs) live on work. those maybe read / written multiple times during a job, so need to be closer to where the work is happening, i.e. in the /work directory

TimothyWillard replied:
I was following the rockfish docs: https://www.arch.jhu.edu/support/storage-and-filesystems/. I don't think the files in consideration here are super large, the conda environment is a bit big relative to the 50GB limit:
```
[twillard@longleaf-login1 ~]$ du -sh /users/t/w/twillard/flepiMoP/
288M	/users/t/w/twillard/flepiMoP/
[twillard@longleaf-login1 ~]$ du -sh /users/t/w/twillard/flepimop-env/
5.7G	/users/t/w/twillard/flepimop-env/
```
I don't think these fit well into the intermediate files that it sounds like scratch is meant for. In your experience have you had any issues with using scratch vs home for `flepiMoP` or the conda environment? Still new to the rockfish setup.

jcblemai replied:
I would put everything in /scratch/ on rockfish (as per the current doc), and in /work/ on longleaf, for both convenience and speed. Home is very small (my home is already 63Gb on longleaf, and around 40Gb in rockfish, most of that conda environments). I think there is some value in having the flepiMoP in the same place especially as we play with several environment and I'd prefer not to have to guess where cd $FLEPI_PATH would take me (and it's easier to rsync the whole thing when needed), but open to have it elsewere.

saraloo replied:
Thanks for these scripts - the install all worked great for me on longleaf. 🎉 

I'm also open to having these things anywhere, but as @jcblemai said I think having everything (including the flepimop libraries) in /work or /scratch makes the most sense, including the flepiMoP folder itself. I understand how installing these in /users or /home would be ideal if flepiMoP was stable but from a practical perspective, I am operationally often changing things within flepimop and reinstalling things run-to-run, playing with my own different environments, jumping between branches, or jumping between different FLEPI_PATH 's (not ideal, but practically this is just what we've had to do with concurrent runs and changes). So for convenience it would be good to just have everything in the same place, imo. 

Separately, with my experience with running stuff in the past I was confused with having to link the specific location of the flepimop-env . I'm fine either way, I just don't think I follow why the change

pearsonca replied:
@saraloo is there a general class of the things you're changing?

saraloo replied:
Responding to both simultaneously. No, i don't think this is normal behavior so feel free to make a judgement call on your end. Just in the past during larger periods of development which inevitably coincide with operational demands I was running two or three different diseases on significantly different gempyor and/or R inference setups from different conda environments (again, don't think this will necessarily be standard, especially now that more people can run stuff). Just flagging that there will be circumstances where flexibility is preferable and want to reduce the possibility of someone setting the wrong flepimop version they;re working on or something, or reducing having to jump around etc to switch branches. 
And sorry, haven't tested the FLEPI_PATH bit yet but that makes sense and I don't anticipate any issues there with setting that.

jcblemai replied:
As we move to the new workflow carl described today, it means that we will have custom branch for runs so you can envision someone running Flu and RSV from the same account but using two different flepiMoP branch. I however think this flexibility can be added alter with the pre-runs scripts that are mentionned below.

Sometime also when running to many parallel run we can have some filesystem lock on the packages, which is always annoying, but I would not worry about it too much.

RE to sara's questions: do we need to specify the emplacement of the conda environement ? 

> . @jcblemai what are the practical differences between the two? My interpretation was that /work was mean for high IO short term storage for active work whereas /users is designed for longer term lower IO (read okay?) storage for libraries/codebases.

This is correct, but flepiMoP does not support writing to other folder other than the project one, so we work from work.

pearsonca commented on build/hpc_install.sh:
```diff
@@ -0,0 +1,157 @@
+fi
+
+# Install the local R packages
+INSTALL_R=$( mktemp )
+cat << EOF > $INSTALL_R
+library(devtools)
+devtools::install_github('HopkinsIDD/flepiMoP', subdir='flepimop/R_packages/flepicommon')
+devtools::install_github('HopkinsIDD/flepiMoP', subdir='flepimop/R_packages/flepiconfig')
+devtools::install_github('HopkinsIDD/flepiMoP', subdir='flepimop/R_packages/inference')
+EOF
+Rscript $INSTALL_R
+rm $INSTALL_R
```
per discussion, can leverage build/setup.R here probably

pearsonca reviewed (changes requested):
Looks generally good, but a few questions to address.

jcblemai commented:
That's really great thank you, ideally we would have a per-cluster specific configuration file that would populate some variables like:
1. where is project_dir
2. where would the final files go
3. what are the pre-processing steps needed before the base flepiMoP one.
Then some cluster-agnostic script would go.

Bit 3. will be used also by the runner script.

We decided to do different commands on the doc instead of a script so that the error are not silent and gradually reported. I would make sure the script exit on failure maybe using `set -e` and `set -x`.

pearsonca commented on build/hpc_install.sh:
```diff
@@ -0,0 +1,157 @@
+rm $INSTALL_R
+
+# Set correct env vars
+export FLEPI_STOCHASTIC_RUN=false
+export FLEPI_RESET_CHIMERICS=TRUE
+export TODAY=`date --rfc-3339='date'`
+
+echo -n "Please set a project path (relative to '$USERDIR'): "
+read PROJECT_PATH
+export PROJECT_PATH="$USERDIR/$PROJECT_PATH"
+
+echo -n "Please set a config path (relative to '$PROJECT_PATH'): "
+read CONFIG_PATH
+export CONFIG_PATH="$PROJECT_PATH/$CONFIG_PATH"
+
+echo -n "Please set a validation date (today is $TODAY): "
+read VALIDATION_DATE
+
+echo -n "Please set a resume location: "
+read RESUME_LOCATION
+
+echo -n "Please set a flepi run index: "
+read FLEPI_RUN_INDEX
```
per @jcblemai discussion: this is a run-specific set of variables, rather than flepimop install level variables.

TimothyWillard replied:
Now handled by `build/flepi_init.sh`.

jcblemai commented on build/hpc_install.sh:
```diff
@@ -0,0 +1,157 @@
+\`\`\`
+Just make sure to \`rm -r model_output\` after running.
+
+Otherwise make sure this diagnostic info looks correct before continuing:
```
This is amazing, would be nice if one could get all this information from a separate command we could ask to be run when there is an issue

TimothyWillard replied:
I've now moved this into `build/flepi_init.sh`. While this also sets some of the key env vars, it also does this sanity checking separate from the install. Does that satisfy the issue here?

jcblemai replied:
yes, thanks

jcblemai commented on build/hpc_install.sh:
```diff
@@ -0,0 +1,157 @@
+# This packages are probably missing from the DESCRIPTION of the R packages
+- r-optparse
+- pip:
+    - git+https://github.com/HopkinsIDD/flepiMoP.git#subdirectory=flepimop/gempyor_pkg
```
Most of the time, folks would want to use a specific branch here. Perhaps the install flepiMoP step should be separated from the build flepiMoP environment step ?

pearsonca replied:
"Separate branches for install vs build for simulation" ... is that what I'm understanding here? That's very... not-a-library energy. I understand dev is still underway, and so we need to make sure the update element of this works well.

jcblemai replied:
yeah, I mean it's a library (you can still run a old config prefectly) but there are also new features constantly needed as our needs increase (there were variant, then age, then seasonality, now races, with harder and harder inference problems and all) so I do think that our scenario modeling hub runs will often require some new features.

TimothyWillard replied:
This now lets you install a consistent set of `gempyor` and R packages from a branch, just checkout the correct branch and then run the install script. Does that satisfy the issue raised here?

jcblemai replied:
yes, thanks !

TimothyWillard commented on build/hpc_install.sh:
```diff
@@ -0,0 +1,157 @@
+- conda-forge
+- defaults
+dependencies:
+- python=3.10
```
As far as I'm aware 3.10 is the only officially supported python version for `gempyor`, see: https://github.com/HopkinsIDD/flepiMoP/blob/427082ce68d68ad8e31c50a537f900aa986b2727/.github/workflows/ci.yml. I've also had great success using 3.11 locally, but that's just local. There are some issues with jumping to 3.12, see: GH-244, maybe we should bump up on priority with 3.13 coming out next week so we can get up to speed with the python ecosystem? I can add 3.11 to our supported versions in either GH-280 or GH-282 if that's of interest, but that's a separate from this PR. In the longer term, our python actions should look more like: https://github.com/ACCIDDA/flepiscripts/blob/initial-package/.github/workflows/ci.yml.

twallema replied:
From experience with bumping up python versions in software: after a python version is released it will take some time for the devs of your downstream dependencies to port to the new python release. Idk how long this takes, but I lag my software about one year behind the new python version.

jcblemai replied:
Love the work with the GitHub action. As it is very painful to get everyone to update their running environment, I would jump ahead and set-up this script on python 3.12 (if the blockers are just deprecation warnings at the moment) or python 3.11 (if worse blocker than that). I realized my rockfish env is 3.12 so I have tested it on this version. 

That only apply if this script is to replace the way one installs on longleaf/rockfish.

TimothyWillard replied:
> As it is very painful to get everyone to update their running environment...

This script should alleviate those issues, and besides, how often do we realistically need to update an environment, once a quarter? That's not very frequent.

I'm happy to bump this up to 3.11 after GH-280 is merged and I'd be willing to do the same for 3.12 in GH-280 (if you would like me to do this then please place a comment there). However, I really don't want us to continue the practice of ignoring warnings so I would not recommend making the default here 3.12. We can reprioritize GH-244 if you'd like to get us up to 3.12.

jcblemai commented:
Also because the above comments focuses on what's missing but it's really awesome that this runs on longleaf (and would have been very useful right now, except that I'm running emcee).

pearsonca commented on flepimop/R_packages/inference/R/install_cli.R:
```diff
@@ -0,0 +1,21 @@
+#' 
+#' @export
+install_cli <- function(
+  path = if (.Platform$OS.type == "unix") normalizePath(file.path("/usr", "local", "bin")) else stop("Unsupported OS")
```
```suggestion
  path = {
    condapth <- Sys.getenv("CONDA_PREFIX")
    if (condapth != "") { 
      file.path(condapth, "bin")
    } else {
      stop("only support default path installation when conda is running.")
    }
  }
```

pearsonca commented on build/setup.R:
```diff
@@ -0,0 +1,36 @@
+
+# TODO sniff for upgrade mode
+
+if (!require(devtools)) {
+  install.packages("devtools", repos = c(getOption("repos"), "http://cran.r-project.org"));
+  stopifnot("Could not load `devtools` package." = require(devtools))
+}
```
why the swap to devtools here? remotes is the requirement, and devtools depends on it (and is just a lightweight pass thru).

pearsonca replied:
to clarify: i'm not sure where devtools dependencies sneak in throughout flepimop, but here its just remotes. if we can eliminate devtools depends elsewhere (e.g. by spinning out the R packages), then we can switch the conda dependency to just remotes, which is like ~20x reduction in dependencies (if they aren't also required by other elements).

TimothyWillard replied:
I think this comment is outdated now. Almost all (minus 1, see GH-332) of the R dependency management is handled directly through conda.

pearsonca commented on build/setup.R:
```diff
@@ -0,0 +1,36 @@
+}
+
+# other dependencies for analysis scripts
+for (pkg in c("ggfortify", "flextable", "optparse", "cowplot")) {
+  if (!requireNamespace(pkg, quietly = TRUE)) { 
+    install.packages(pkg, repos = c(getOption("repos"), "http://cran.r-project.org"));
+    if (!requireNamespace(pkg)) {
+       stop(sprintf("Could not install and/or load `%s` package.", pkg))
+    }
+  }
+}
```
these can all be moved into the conda env specification (all just the stand `r-...` names)

pearsonca replied:
(agree optparse could go into inference as a depends; makes more sense as a suggests, but there isn't really a way to differentiate "use" vs "develop" suggests, and the "use" environment doesn't need testthat)

TimothyWillard replied:
I think this comment is outdated now. Almost all (minus 1, see https://github.com/HopkinsIDD/flepiMoP/issues/332) of the R dependency management is handled directly through conda. I moved `optparse` into the imports of the `inference` package since that is required to install the CLI tools exposed by that package.

TimothyWillard commented on build/hpc_install.sh:
```diff
@@ -0,0 +1,157 @@
+# Ensure we have a $FLEPI_PATH
+if [ -z "${FLEPI_PATH}" ]; then
+    echo "An explicit \$FLEPI_PATH was not provided, setting to '$USERDIR/flepiMoP'."
+    export FLEPI_PATH="$USERDIR/flepiMoP"
```
Now handled by `build/flepi_init.sh`.

TimothyWillard commented on build/hpc_install.sh:
```diff
@@ -0,0 +1,157 @@
+Rscript $INSTALL_R
+rm $INSTALL_R
+
+# Set correct env vars
```
Now handled by `build/flepi_init.sh`.

TimothyWillard commented on flepimop/R_packages/inference/R/install_cli.R:
```diff
@@ -0,0 +1,32 @@
+  )
+  from <- scriptfiles
+  to <- file.path(path, gsub("\\.R$", "", basename(scriptfiles)))
+  to_remove <- file.exists(to)
+  if (length(to_remove)) {
+    file.remove(to)
+  }
```
Added this to get rid of a warning about the file existing on reinstall that looks like this:
```R
> library(inference); inference::install_cli()
[1] FALSE FALSE
Warning messages:
1: In file.symlink(from, to) :
  cannot symlink '/users/t/w/twillard/flepimop-env/lib/R/library/inference/scripts/flepimop-inference-main.R' to '/users/t/w/twillard/flepimop-env/bin/flepimop-inference-main', reason 'File exists'
2: In file.symlink(from, to) :
  cannot symlink '/users/t/w/twillard/flepimop-env/lib/R/library/inference/scripts/flepimop-inference-slot.R' to '/users/t/w/twillard/flepimop-env/bin/flepimop-inference-slot', reason 'File exists'
```
Are there side effects from doing this that I'm missing @pearsonca?

pearsonca replied:
this should be fine. if we don't change the names of the target files, we actually don't need to rewrite the symlinks when updating the file.

but i'm iffy enough on that if that it seems best to always blowaway and recreate the symlink

TimothyWillard replied:
Alternative is `suppressWarnings(file.symlink(from, to))` which feels heavy handed and leaves the door open to suppressing an actual warning though.

TimothyWillard commented:
This is ready for review again. One caveat I have is that I think it might be easier to merge GH-323 first, that might clean up/clarify some of the changes here. Re some of the issues raised in the course of this PR:

> That's really great thank you, ideally we would have a per-cluster specific configuration file that would populate some variables like:

Yeah, a repository of per cluster specific configurations would be really nice, maybe not necessarily yaml for ease of parsing by bash. Do you want to do that now? I think this requires setting up a public repo under ACCIDDA with these configurations and that would all have to be done before I could experiment integrating that here.

> gempyor benefits greatly from the performance improvments of 3.11, any reasons to pin the python version here ?

I think this is addressed in GH-280, once that is merged I would have few issues with changing this.

Re: https://github.com/HopkinsIDD/flepiMoP/pull/329#discussion_r1785089348

I don't think we came to a good conclusion on what to do about directory placement on rockfish. Changing this code wise is trivial, so just need to make a decision. Any additional thoughts in light of that thread @pearsonca, @fang19911030, or @shauntruelove?

Lastly, I need to make some edits to the documentation about how to use this. So while the code is "done" this PR (pending review) does not totally resolve the HPC install issue. I will open a separate PR into the `documentation-gitbook` branch and reference it to this PR.

pearsonca commented on build/flepi_init.sh:
```diff
@@ -0,0 +1,129 @@
+if [ ! -d "$FLEPI_PATH" ]; then
+    echo "You must have a clone of flepiMoP at $FLEPI_PATH to use this script."
+    exit 1
+elif [ ! -d "$FLEPI_PATH/.git" ]; then
+    echo "The flepiMoP found at '$FLEPI_PATH' is not a git clone, unsure of how to proceed."
+    exit 1
+fi
```
(also, should have noticed this earlier)

TimothyWillard replied:
Yeah, this shouldn't be in `flepi_init.sh`. Removed.

TimothyWillard commented on build/flepi_init.sh:
```diff
@@ -0,0 +1,129 @@
+
+echo -n "Please set a config path (relative to '$PROJECT_PATH'): "
+read CONFIG_PATH
+export CONFIG_PATH="$PROJECT_PATH/$CONFIG_PATH"
```
Added a warning, users might want to clone the project later or use bogus directories/files like I am in testing these scripts.

TimothyWillard commented on build/hpc_install.sh:
```diff
@@ -0,0 +1,125 @@
+if [ ! -d "$FLEPI_PATH" ]; then
+    echo "Did not find flepiMoP at '$FLEPI_PATH', cloning on your behalf."
+    git clone git@github.com:HopkinsIDD/flepiMoP.git $FLEPI_PATH
+elif [ ! -d "$FLEPI_PATH/.git" ]; then
```
Added.

TimothyWillard commented on build/hpc_install.sh:
```diff
@@ -0,0 +1,125 @@
+dependencies <- dependencies[!grepl("^R(\\\\(.*\\\\))?$", dependencies)]
+# Print the dependencies
+cat(paste(dependencies, collapse = ","))
+EOF
```
Yeah, this logic has been moved into `build/create_environment_yml.R` and automated with `.github/workflows/conda-env.yml`.

TimothyWillard commented on build/hpc_install.sh:
```diff
@@ -0,0 +1,125 @@
+# Print the dependencies
+cat(paste(dependencies, collapse = ","))
+EOF
+R_DEPENDS=$( Rscript $R_DEPENDS_SCRIPT $FLEPI_PATH )
```
Outside of, but it doesn't matter either way. Script only uses base R functionality so can use any R install. Also removed.

jcblemai commented:
Hey, I added in some comments, just on some specific points:
> Yeah, a repository of per cluster specific configurations would be really nice, maybe not necessarily yaml for ease of parsing by bash. Do you want to do that now? 

no, not in this PR. I think we may even as this inside this flepiMoP repository for convenience, or as a shared maintained folder on each cluster.

Is this script fail-proof enough that it can replace the installation guide ? Or should one still copy command to command ?

TimothyWillard commented:
> no, not in this PR. I think we may even as this inside this flepiMoP repository for convenience, or as a shared maintained folder on each cluster.

I think this repo is the wrong place and placing in a shared folder on each cluster defeats the purpose of having said config (need to know something about the directory structure on each cluster to access the config for each cluster). Can hash out the rest on GH-342.

> Is this script fail-proof enough that it can replace the installation guide ? Or should one still copy command to command ?

I'm not sure what the bar for "fail-proof enough" is here, but I plan on replacing the current HPC install/update guides in the flepiMoP wiki in a separate PR into the `documentation-gitbook` branch and reference it to this PR. I could do that now, but have a slight preference for just doing after in case the scripts change again.

pearsonca commented:
> I'm not sure what the bar for "fail-proof enough" is here, but I plan on replacing the current HPC install/update guides in the flepiMoP wiki in a separate PR into the `documentation-gitbook` branch and reference it to this PR. I could do that now, but have a slight preference for just doing after in case the scripts change again.

Agree, separate PR to docs - its not like this makes those directions *not* work.

pearsonca reviewed (approved):
lgtm.

I have one slight concern re the flepi_init.sh: I don't think those variables are *required* to be set, and that doesn't exactly come across in the script. Probably harmless to do so.

However, I hope in the middle term most of the environmental variable initialization will become obsolete. Using the environmental variables as we do strikes me as too-much-magic, and that their advantages are easily realized via alternative approaches, e.g. instead of

```bash
CONFIG_FILE=whatever
flepimop simulate
```
one could do
```bash
CONFIG_FILE=whatever
flepimop simulate -c $CONFIG_FILE
```
which surfaces precisely what's going on while still allowing flexibility, and isn't that many more characters to type.

TimothyWillard commented:
This is ready for review again. Some notable changes since the last time this was looked at:

1. The default python version was upped from 3.10 to 3.11.
2. The directory structure has changed per this discussion: https://github.com/HopkinsIDD/flepiMoP/pull/329#discussion_r1785089348.
3. The flepiMoP path and conda environment have defaults, but are configurable with prompts for more dynamic use by operations.

pearsonca reviewed (dismissed):
looks good to me

jcblemai commented:
Looks good to me, provided a small batch run works with it. Thanks for the great work. I like installing with pip -e because it means I can update gempyor very fast/switch branch and have the new thing ready.

jcblemai commented:
Your first comment about how to use this PR should be added to the doc right ? and replace manual environment creation.

TimothyWillard commented:
> I like installing with pip -e because it means I can update gempyor very fast/switch branch and have the new thing ready.

This has been added now.

> Your first comment about how to use this PR should be added to the doc right ? and replace manual environment creation.

The first comment is a bit out of date now, usage has changed some. However, I plan on replacing the current HPC install/update guides in the flepiMoP wiki in a separate PR into the `documentation-gitbook` branch and reference it to this PR.

Conda environment management has changed some from the discussion in https://github.com/HopkinsIDD/flepiMoP/pull/329#discussion_r1785089348 per an slack discussion with @jcblemai and @saraloo. Will now use a default environment in `~/.conda` rather than an absolute path in the work directories of the clusters.

pearsonca commented:
Not sure where @jcblemai's comment went re branches wanted for different operational runs, but working trees seem like an option: https://stackoverflow.com/questions/2048470/git-working-on-two-branches-simultaneously - might be something to do via init script?

TimothyWillard commented:
> Not sure where @jcblemai's comment went re branches wanted for different operational runs, but working trees seem like an option: https://stackoverflow.com/questions/2048470/git-working-on-two-branches-simultaneously - might be something to do via init script?

This is currently handled in `batch/inference_job_launcher.py` which is I think where you want to handle branching for job submission and hasn't been touched by this PR yet. Do you want me to make changes to that in this PR as well now? Maybe this is something better fit for https://github.com/ACCIDDA/flepiscripts/pull/1?

> Looks good to me, provided a small batch run works with it.

After the recent round of edits I was able to submit one of the recent Flu configs to rockfish using an environment setup and initialized using the `build/hpc_install_or_update.sh` and `batch/hpc_init.sh` scripts. It was a small submission (2 jobs, 200 iterations) so the actual outputs aren't helpful, but does demonstrate that the tools provided in this PR can setup a HPC environment suitable for batch submission.

pearsonca commented:
> This is currently handled in `batch/inference_job_launcher.py` which is I think where you want to handle branching for job submission and hasn't been touched by this PR yet. Do you want me to make changes to that in this PR as well now? Maybe this is something better fit for [ACCIDDA/flepiscripts#1](https://github.com/ACCIDDA/flepiscripts/pull/1)?

Let's aim for "next up" on that - I'd like an integration of these scripts w/ the CLI to support `flepimop update` and `flepimop init ...` actions.

TimothyWillard commented:
> Let's aim for "next up" on that - I'd like an integration of these scripts w/ the CLI to support flepimop update and flepimop init ... actions.

Sure, can you create an issue for that then to move discussion of details there?
